File: question/leetcode/leetcode0142.Linked-List-Cycle-II/leetcode0142.py
def detect_cycle(head):
    if not head or not head.next:  # 如果链表为空或只有一个节点，则没有环
        return None

    fast = slow = head
    has_circle = False
    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next
        if fast == slow:
            has_circle = True
            break
    if not has_circle:
        return None
    slow = head
    while slow != fast:
        fast = fast.next
        slow = slow.next
    return slow

File: question/leetcode/leetcode0142.Linked-List-Cycle-II/leetcode0142.go
package leetcode

import "clarencedu/algorithm-go/structure"

type ListNode = structure.ListNode

func detectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}
	hasCircle := false
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
		if fast == slow {
			hasCircle = true
			break
		}
	}
	if !hasCircle {
		return nil
	}
	slow = head
	for slow != fast {
		fast = fast.Next
		slow = slow.Next
	}
	return slow
}